PyNEST: pep8 compliance and Numpy-formatted docstrings.

flinz opened the pull request:
This PR came about due to https://github.com/nest/nest-simulator/issues/188 and https://github.com/nest/nest-simulator/issues/257.

# Major changes:
- added numpy standardized docstring ([see here for specs](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt)) to all functions in ``pynest/nest``, excluding ``pynest/nest/tests`` in preparation for autogeneration of documentation
- pep8 compliant formatting for almost all python files in ``pynest/nest``, so the nest directory now checks out as almost clean under pep8 (see issues below)
- ``raster_plot.py`` has been slightly rewritten to simplify redundant keyword arguments

# Remaining issues
- I tried as best I could to infer missing parameter types etc for the doc-strings, there might be some slight "mistypings".
- Problem with ``pynest/nest/__init__.py`` scipy bugfix: Right now pep8 checks out as almost clean, except the 4 errors below. This is due to the fact that the [imports of ``pynestkernel`` and ``hl_api_helper``](https://github.com/nest/nest-simulator/pull/263/files#diff-14ca38d825613569e5bb0ab8ea82e266R58) come after the scipy bugfix ([see here](https://github.com/nest/nest-simulator/pull/263/files#diff-14ca38d825613569e5bb0ab8ea82e266L37)). Does anybody know more about this scipy import issue? Can we move the 4 import lines to the top of the file? I do not see scipy being imported anywhere there, so in my opinion this should be safe to do.
```
% pep8 nest
nest/__init__.py:57:1: E402 module level import not at top of file
nest/__init__.py:58:1: E402 module level import not at top of file
nest/__init__.py:60:1: E402 module level import not at top of file
nest/__init__.py:61:1: E402 module level import not at top of file
```

heplesser commented:
@flinz Thanks for your work! At the moment, the checks fail due to some copyright header problem---could you check that?

Import order, especially wrt SciPy/NumPy has been a problem for a long time, and one sensitive to a number of details, if I remember correctly. I would tend not to fiddle with the order of imports just to make PEP8 happy, in particular in an `__init__.py` that is not part of the public Python interface of NEST. I think this is a place where we probably should suppress the PEP8 warning.

flinz commented:
@heplesser copyright header errors should be fixed now.
I also moved all imports in ``__init__.py`` back to the original order (I had already moved two) just to be sure.

jougs commented on pynest/nest/__init__.py:
```diff
@@ -82,27 +87,42 @@ def catching_sli_run(cmd):
+    """Convenience function for executing the sequence
+    sli_push(args); sli_run(s); y=sli_pop().
+
+    The function takes an
```
Something is missing here. Probably, the sentence can just be deleted, as the thing with the arbitrary number of arguments is written below.

jougs commented on pynest/nest/lib/hl_api_connections.py:
```diff
@@ -27,44 +27,84 @@
+    synapse_model : str, optional
+        Only connections with this synapse model are returned
+    synapse_type : str, optional
+        Only connections with this synapse type are returned
```
I'd add the fact that this is the same as `synapse_model`. On the other side, the function is deprecated since long and can as well be deleted right after this PR is merged.

jougs replied:
Forget this! I just saw that it is mentioned below. Nevermind.

heplesser commented:
:+1: from my side.

jougs commented:
I gave up reviewing after I've read half of it and did not find a single error. So, pragmatic :+1: from my side as this is definitely better than what we had before. @tammoippen or @abigailm: if you agree with my lax style of reviewing, could you please merge?
